grype/db/curator_test.go

```diff
@@ -364,3 +373,77 @@ func TestCurator_validateStaleness(t *testing.T) {
 		})
 	}
 }
+
+func TestCuratorTimeoutBehavior(t *testing.T) {
+	failAfter := 10 * time.Second
+	success := make(chan struct{})
+	errs := make(chan error)
+	timeout := time.After(failAfter)
+
+	hangForeverHandler := func(w http.ResponseWriter, r *http.Request) {
+		select {} // hang forever
+	}
+
+	ts := httptest.NewServer(http.HandlerFunc(hangForeverHandler))
+
+	cfg := Config{
+		DBRootDir:           "",
+		ListingURL:          fmt.Sprintf("%s/listing.json", ts.URL),
+		CACert:              "",
+		ValidateByHashOnGet: false,
+		ValidateAge:         false,
+		MaxAllowedBuiltAge:  0,
+		ListingFileTimeout:  400 * time.Millisecond,
+		UpdateTimeout:       400 * time.Millisecond,
+	}
+
+	curator, err := NewCurator(cfg)
+	require.NoError(t, err)
+
+	u, err := url.Parse(fmt.Sprintf("%s/some-db.tar.gz", ts.URL))
+	require.NoError(t, err)
+
+	entry := ListingEntry{
+		Built:    time.Now(),
+		Version:  5,
+		URL:      u,
+		Checksum: "83b52a2aa6aff35d208520f40dd36144",
+	}
+
+	downloadProgress := progress.NewManual(10)
+	importProgress := progress.NewManual(10)
+	stage := progress.NewAtomicStage("some-stage")
+
+	runTheTest := func(success chan struct{}, errs chan error) {
+		_, _, _, err = curator.IsUpdateAvailable()
+		if err == nil {
+			errs <- errors.New("expected timeout error but got nil")
+			return
+		}
+		if !strings.Contains(err.Error(), "Timeout exceeded") {
+			errs <- fmt.Errorf("expected %q but got %q", "Timeout exceeded", err.Error())
+			return
+		}
+
+		err = curator.UpdateTo(&entry, downloadProgress, importProgress, stage)
+		if err == nil {
+			errs <- errors.New("expected timeout error but got nil")
+			return
+		}
+		if !strings.Contains(err.Error(), "Timeout exceeded") {
+			errs <- fmt.Errorf("expected %q but got %q", "Timeout exceeded", err.Error())
+			return
+		}
+		success <- struct{}{}
+	}
+	go runTheTest(success, errs)
+
+	select {
+	case <-success:
+		return
+	case err := <-errs:
+		t.Error(err)
+	case <-timeout:
```

Comment:
Since we're doing a test that combines the calls of `IsUpdateAvailable` and `UpdateTo` is there a chance this can race incorrectly? 

I guess the gap is so large from the configured timeouts and `failAfter` (more than 10x) that it doesn't matter.

LGTM

Reply:
@spiffcs I wanted to make sure I understand what you mean by "race incorrectly." I think you're saying that since (1) the test fails on a timeout and (2) the assertion expects a different timeout, we could get the wrong timeout. Is that correct?

I think that since we're effectively racing a 0.8 second timer against a 10 second timer, the test should be pretty stable. I put in the 10 second timer because otherwise removing one of the timeouts causes the test suite to hang for 10 minutes while the overall testing timer runs out, which seems like worse behavior.

Please let me know if I've addressed these concerns adequately and then I'll merge.